modules/apps/portal-crypto-hash/portal-crypto-hash-test/src/testIntegration/java/com/liferay/portal/crypto/hash/processor/test/HashProcessorTest.java

```diff
@@ -47,51 +48,85 @@
+			hashContextBuilder1.createHashGenerationContextBuilder();
+
+		defaultSaltHashGenerationContext1 = hashGenerationContextBuilder1.build(
+			SaltCommand.generateDefaultSizeSalt());
+
+		variableSaltHashGenerationContext1 =
+			hashGenerationContextBuilder1.build(
+				SaltCommand.generateVariableSizeSalt(_VARIABLE_SIZE));
+
+		notSaltHashGenerationContext1 = hashGenerationContextBuilder1.build();
+
+		HashContextBuilder hashContextBuilder2 =
+			_hashProcessor.createHashContextBuilder(_MESSAGE_DIGEST_ALGO_2);
+
+		HashGenerationContext.Builder hashGenerationContextBuilder2 =
+			hashContextBuilder2.createHashGenerationContextBuilder();
+
+		defaultSaltHashGenerationContext2 = hashGenerationContextBuilder2.build(
+			SaltCommand.generateDefaultSizeSalt());
+
+		variableSaltHashGenerationContext2 =
+			hashGenerationContextBuilder2.build(
+				SaltCommand.generateVariableSizeSalt(_VARIABLE_SIZE));
+	}
+
 	@Test
 	public void testHashGenerationAndHashVerificationWithMultipleContexts()
 		throws Exception {
 
-		PepperContextBuilder pepperContextBuilder1 =
-			_hashProcessor.createHashContextBuilder(_MESSAGE_DIGEST_ALGO_1);
-
 		HashGenerationResponse hashGenerationResponse1 =
 			_hashProcessor.generate(
-				_PASSWORD.getBytes(),
-				pepperContextBuilder1.pepperProvider(
-					_PEPPER.getBytes()
-				).buildGenerationContext(
-					SaltCommand.generateDefaultSizeSalt()
-				));
+				_PASSWORD.getBytes(), defaultSaltHashGenerationContext1);
 
-		Optional<byte[]> optionalSalt1 = hashGenerationResponse1.getSalt();
+		HashFlavor hashFlavor1 = hashGenerationResponse1.getHashFlavor();
 
-		final PepperContextBuilder pepperContextBuilder2 =
-			_hashProcessor.createHashContextBuilder(_MESSAGE_DIGEST_ALGO_2);
+		Optional<String> optionalPepperId1 = hashFlavor1.getPepperId();
+		Optional<byte[]> optionalSalt1 = hashFlavor1.getSalt();
 
 		HashGenerationResponse hashGenerationResponse2 =
 			_hashProcessor.generate(
 				hashGenerationResponse1.getHash(),
-				pepperContextBuilder2.pepperProvider(
-					_PEPPER.getBytes()
-				).buildGenerationContext(
-					SaltCommand.generateVariableSizeSalt(_VARIABLE_SIZE)
-				));
+				variableSaltHashGenerationContext2);
+
+		HashFlavor hashFlavor2 = hashGenerationResponse2.getHashFlavor();
 
-		Optional<byte[]> optionalSalt2 = hashGenerationResponse2.getSalt();
+		Optional<String> optionalPepperId2 = hashFlavor2.getPepperId();
+		Optional<byte[]> optionalSalt2 = hashFlavor2.getSalt();
+
+		HashContextBuilder hashContextBuilder1 =
+			_hashProcessor.createHashContextBuilder(_MESSAGE_DIGEST_ALGO_1);
+
+		HashVerificationContext.Builder hashVerificationContextBuilder1 =
+			hashContextBuilder1.createHashVerificationContextBuilder();
 
 		HashVerificationContext hashVerificationContext1 =
-			pepperContextBuilder1.pepperProvider(
-				_PEPPER.getBytes()
-			).buildVerificationContext(
-				optionalSalt1.get()
-			);
+			hashVerificationContextBuilder1.pepper(
+				optionalPepperId1.orElse("")
+			).salt(
```

Comment:
hey Arthur, 
if we set the pepper here then it is not getting it from the pepper storage, is it? 
this should be in sync with the strategy to get a new pepper when we generate a new hash, right?

Reply:
Hi @csierra ,
I am not quite following your question. Here what is passed is a pepperId(supposely from a db), and the HashProcessor will be responsible for getting the actual pepper from pepper storage **using** this pepperId.

Yes, I think it is synced with the strategy, because the HashGeneration will give a pepperId to the user, and the user will store the pepperId to the db. Then when in a verifiation, the user will need to provide the pepperId to the processor.

Reply:
ah... I misread it... I though it was passing the pepper directly, sorry for that...